src/s3ql/backends/gdrive.py

```diff
@@ -0,0 +1,599 @@
+from .common import (AbstractBackend, DanglingStorageURLError, NoSuchObject,retry,retry_generator,AuthenticationError)
+import apiclient.discovery
+import apiclient.http
+import time
+import base64
+import hashlib
+import httplib2
+import ssl
+import http
+import socket
+import io
+import json
+import oauth2client.client
+import tempfile
+
+log = logging.getLogger(__name__)
+logging.getLogger("googleapiclient.discovery").setLevel(logging.WARNING)
+
+
+
+# Maximum number of keys that can be deleted at once
+MAX_KEYS = 1000
+credentials= None
+mainFolder= None
+class Backend(AbstractBackend, metaclass=ABCDocstMeta):
+    '''
+    A backend that stores data in Google Drive.
+    '''
+
+    known_options = set()
+    MIME_TYPE_BINARY = 'application/octet-stream'
+    MIME_TYPE_FOLDER = 'application/vnd.google-apps.folder'
+    TOKEN_URI = 'https://accounts.google.com/o/oauth2/token'
+    USER_AGENT = 'S3QL-Backend'
+
+    MAX_PROPERTY_SIZE = 124
+    MAX_REF_CHUNK_SIZE = 100
+    VALUE_BASE64 = 'b64:'
+    VALUE_REF = 'ref:'
+    
+
+    def __init__(self, storage_url, login, password, options):
+        '''Initialize local backend
+        Login and password are ignored.
+        '''
+        # Unused argument
+        #pylint: disable=W0613
+
+        super().__init__()
+        self.login = login
+        self.password = password
+        client_secret_and_refresh_token = self.password.split(':')
+        if len(client_secret_and_refresh_token) != 2:
+            raise AuthenticationError("Invalid Password format, must be  secretApplication:refreshToken")
+        self.client_secret = client_secret_and_refresh_token[0]
+        self.refresh_token = client_secret_and_refresh_token[1]
+        self.options = options
+        self.delete_map_state = dict()
+
+        # get google drive service
+        global credentials        
+        global mainFolder
+        if credentials == None:
+            credentials = self._get_credentials()
+        http = credentials.authorize(httplib2.Http())        
+
+        self.service = apiclient.discovery.build('drive', 'v3', http=http,cache_discovery=False)
+
+        # get gdrive folder
+        if mainFolder==None:
+            folderPath = storage_url[len('gdrive://'):].rstrip('/')
+            self.folder = self._lookup_file(folderPath)
+            if self.folder != None:
+                if self.folder['mimeType'] != Backend.MIME_TYPE_FOLDER:
+                    raise DanglingStorageURLError(folderPath)
+            else:
+                self.folder = self.service.files().create(body={'name' : folderPath,'mimeType' : Backend.MIME_TYPE_FOLDER},fields='id').execute()
+            mainFolder=self.folder
+        else:
+            self.folder = mainFolder
+
+    def _get_credentials(self):
+        return oauth2client.client.OAuth2Credentials(access_token="1234",client_id=self.login,client_secret=self.client_secret,
```

Comment:
Please keep lines under 100 characters (not just here but everywhere in the code)